File: services/server/src/safe_socket/safe_socket.py
import socket

# TODO: Complete with a short-read/short-write tolerant implementation


def recv_all(socket: socket.socket, size):
    buffer = bytearray(size)
    received = 0

    while received < size:
        try:
            chunk = socket.recv(size - received)
            if not chunk:
                raise ConnectionError("connection closed while receiving data")
        except OSError:
            raise

        buffer[received:received + len(chunk)] = chunk
        received += len(chunk)

    return bytes(buffer)

def send_all(socket: socket.socket, bytes):
    remaining = memoryview(bytes)

    while len(remaining) > 0:
        try:
            sent = socket.send(remaining)
        except OSError:
            raise

        if sent == 0:
            continue

        remaining = remaining[sent:]

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
	"io"
)

func SendAll(socket io.Writer, bytes []byte) error {

	for len(bytes) > 0 {
		n, err := socket.Write(bytes)
		if err != nil {
			return err
		}
		if n == 0 {
			continue
		}
		bytes = bytes[n:]
	}

	return nil

}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	received := 0
	for received < size {
		n, err := socket.Read(buff[received:])
		received += n
		if err != nil {
			if received == size {
				return buff, nil
			}
			return buff[:received], err
		}
		if n == 0 {
			continue
		}
	}

	return buff, nil
}
